test/suite/test_hs18.py

```diff
@@ -0,0 +1,415 @@
+
+# test_hs18.py
+# Test various older reader scenarios
+class test_hs18(wttest.WiredTigerTestCase):
+    conn_config = 'cache_size=5MB,eviction=(threads_max=1)'
+    session_config = 'isolation=snapshot'
+
+    def check_value(self, cursor, value):
+        cursor.set_key(str(0))
+        self.assertEqual(cursor.search(), 0)
+        self.assertEqual(cursor.get_value(), value)
+        cursor.reset()
+
+    def start_txn(self, sessions, cursors, values, i):
+        # Start a transaction that will see update 0.
+        sessions[i].begin_transaction()
+        self.check_value(cursors[i], values[i])
+
+    def test_hs18(self):
+        uri = 'table:test_hs18'
+        self.session.create(uri, 'key_format=S,value_format=S')
+        session2 = self.setUpSessionOpen(self.conn)
+        cursor = self.session.open_cursor(uri)
+        cursor2 = session2.open_cursor(uri)
+
+        value0 = 'f' * 500
+        value1 = 'a' * 500
+        value2 = 'b' * 500
+        value3 = 'c' * 500
+        value4 = 'd' * 500
+        value5 = 'e' * 500
+
+        # Insert an update at timestamp 3
+        self.session.begin_transaction()
+        cursor[str(0)] = value0
+        self.session.commit_transaction('commit_timestamp=' + timestamp_str(3))
+
+        # Start a long running transaction which could see update 0.
+        session2.begin_transaction()
+        self.check_value(cursor2, value0)
+
+        # Insert an update at timestamp 5
+        self.session.begin_transaction()
+        cursor[str(0)] = value1
+        self.session.commit_transaction('commit_timestamp=' + timestamp_str(5))
+
+        # Insert another update at timestamp 10
+        self.session.begin_transaction()
+        cursor[str(0)] = value2
+        self.session.commit_transaction('commit_timestamp=' + timestamp_str(10))
+
+        # Insert a bunch of contents to fill the cache
+        for i in range(2000, 10000):
+            self.session.begin_transaction()
+            cursor[str(i)] = value3
+            self.session.commit_transaction()
+
+        # Commit an update without a timestamp on our original key
+        self.session.begin_transaction()
+        cursor[str(0)] = value4
+        self.session.commit_transaction()
+
+        # Commit an update with timestamp 15
+        self.session.begin_transaction()
+        cursor[str(0)] = value5
+        self.session.commit_transaction('commit_timestamp=' + timestamp_str(15))
+
+        # Check our value is still correct.
+        self.check_value(cursor2, value0)
+
+        # Insert a bunch of other contents to trigger eviction
+        for i in range(10001, 11000):
+            self.session.begin_transaction()
+            cursor[str(i)] = value3
+            self.session.commit_transaction()
+
+        # Check our value is still correct.
+        self.check_value(cursor2, value0)
+
+    # Test that we don't get the wrong value if we read with a timestamp originally.
+    def test_read_timestamp_weirdness(self):
```

Comment:
So we are not going to add the test case that will make the reader seeing newer updates if read with a timestamp?

like U@15 -> U@0 -> (start read) U@10 ->U@5, read timestamp 5

Reply:
I can add that case

Reply:
Added it in latest commit, it works as expected in that we see a newer value.